### ISSUE TITLE
1.16: Should we remove Printf from the API

### ISSUE THREAD

**Issue by whyman** [pupnp/pupnp#365]
It seems like `UpnpLog::UpnpPrintf` is only meant for internal library usage.

The `UpnpLog` class should still be there for configuring the library logging stuff, but we shouldn't expose the actual logging method itself.

Thoughts?

**Comment by Vollstrecker** [comment]
Absolutely right. But as I still didn't get the time to find out what options are needed to only expose the functions that should be, it wouldn't make a big difference, as atm everything is reachable from outside.

**Comment by mrjimenez** [comment]
It could be removed, it is really a convenience to make it easier to do tests. But I think that right now, not even the tests use it explicitly, only indirectly.

**Comment by mrjimenez** [comment]
Thanks for raising this, @whyman — and agreed (@Vollstrecker, @mrjimenez).

To be precise about the current state: `UpnpPrintf` already lacks `UPNP_EXPORT_SPEC`, and CMake builds use `C_VISIBILITY_PRESET "hidden"`, so the symbol is not exported from the shared library in CMake builds. However the situation is inconsistent across build systems:

- **CMake builds**: `UpnpPrintf` is hidden (no `UPNP_EXPORT_SPEC` + `-fvisibility=hidden`); the `test_log.c` calls are inside `#ifdef UPNP_HAVE_DEBUG` blocks that CMake never enters, so they are dead code. `test_parse_uri.c` provides its own stub definition and does not use the library's copy.
- **Autotools `--enable-debug` builds**: no `-fvisibility=hidden` by default, so `UpnpPrintf` *is* exported and `test_log.c` does call the real function.

The function is also still declared in the public header `upnp/inc/upnpdebug.h`, which is misleading — it implies callability that the linker won't satisfy in CMake builds, and creates the platform-dependent inconsistency above.

Proposed fix:

1. Remove the `UpnpPrintf` declaration from `upnp/inc/upnpdebug.h` (the other logging functions — `UpnpInitLog`, `UpnpSetLogLevel`, `UpnpCloseLog`, `UpnpSetLogFileNames`, `UpnpGetDebugFile` — are correctly marked `UPNP_EXPORT_SPEC` and genuinely public).
2. Move the declaration to an internal header (e.g. `upnp/src/api/upnpdebug_internal.h`), so that `upnpdebug.c`, `gena/`, `soap/`, `ssdp/` etc. can still call it.
3. Update `test_log.c` to include the internal header rather than relying on the public one.
4. Optionally add an explicit `__attribute__((visibility("hidden")))` to the definition in `upnpdebug.c` for consistency in autotools builds.

This is a minor API cleanup with no ABI impact for users who were using the function correctly (i.e. not at all — it was never reliably exported). We can include this in the next maintenance release.

**Comment by Vollstrecker** [comment]
> To be precise about the current state: `UpnpPrintf` already lacks `UPNP_EXPORT_SPEC`, and CMake builds use `C_VISIBILITY_PRESET "hidden"`, so the symbol is not exported from the shared library in CMake builds. However the situation is inconsistent across build systems:

CMake sets also [LIBUPNP_EXPORTS`](https://github.com/pupnp/pupnp/blob/1dc103b094b45d72c732eacae3ce8ef3bfa96d8b/upnp/CMakeLists.txt#L127) and [UPNP_STATIC_LIB](https://github.com/pupnp/pupnp/blob/1dc103b094b45d72c732eacae3ce8ef3bfa96d8b/upnp/CMakeLists.txt#L191) which should select the right declspec in [UpnpGlobal.h.](https://github.com/pupnp/pupnp/blob/1dc103b094b45d72c732eacae3ce8ef3bfa96d8b/upnp/inc/UpnpGlobal.h#L26)

>     * **CMake builds**: `UpnpPrintf` is hidden (no `UPNP_EXPORT_SPEC` + `-fvisibility=hidden`); the `test_log.c` calls are inside `#ifdef UPNP_HAVE_DEBUG` blocks that CMake never enters, so they are dead code. `test_parse_uri.c` provides its own stub definition and does not use the library's copy.

UPNP_HAVE_DEBUG is missing. Most other stuff seems to rely on Debug/NDEBUG, maybe this should all be unified.

>     1. Remove the `UpnpPrintf` declaration from `upnp/inc/upnpdebug.h` (the other logging functions — `UpnpInitLog`, `UpnpSetLogLevel`, `UpnpCloseLog`, `UpnpSetLogFileNames`, `UpnpGetDebugFile` — are correctly marked `UPNP_EXPORT_SPEC` and genuinely public).
>
>     2. Move the declaration to an internal header (e.g. `upnp/src/api/upnpdebug_internal.h`), so that `upnpdebug.c`, `gena/`, `soap/`, `ssdp/` etc. can still call it.
> 
>     3. Update `test_log.c` to include the internal header rather than relying on the public one.

If it's not meant to be public, then that's the best option. Maybe that's also a good time to check if all installed headers are meant to be public (or for missing ones), so they can be split in private and public ones.
 
>     4. Optionally add an explicit `__attribute__((visibility("hidden")))` to the definition in `upnpdebug.c` for consistency in autotools builds.

As the DEFAULT_PRESET is set to hidden, it should be hidden already when not marked otherwise.
 
> This is a minor API cleanup with no ABI impact for users who were using the function correctly (i.e. not at all — it was never reliably exported). We can include this in the next maintenance release.

Yep, and the deprecation-warning for the old options was also long enough there that everyone should have noticed. The question there is now "Make it an Error so the last one using it will notice for 2 or 3 releases, or just ignore them silently?".

**Comment by mrjimenez** [comment]
I am thinking about naming the next release 2.0.0 as if it were a real API breakage. So no one complains. Switching to 2.0.0 would be trouble only for those rare cases when someone managed to use UpnpPrintf.

Later today we will see what Claude thinks of your answer, I can't do it right now. 🤣

**Comment by Vollstrecker** [comment]
Yeah, as you can see in #587 version is set to 20.1.0. I guess for this change the math doesn't work out (rough guess, still not understanding what to expect).

And as you can see there we get a log (96) warnings. Most of them are because of unknown pragma GCC, maybe we should drop these and return to classical #if or #ifdef.

**Comment by mrjimenez** [comment]
That is because a previous patch added interface to the library (SSL related).

From what I can see, the autotools and the cmake versions do different things, what a mess.

The best option is to do an API break. It is not **really** an API break, since people are not supposed to use UpnpPrintf, so whoever uses it, just change to printf, no big deal.

**Comment by mrjimenez** [comment]
Draft PR opened: https://github.com/pupnp/pupnp/pull/588

Branch `chore/issue-365-116-should-we-remove-printf-from-the-api` — regression test committed (confirmed failing on unfixed autotools builds) and fix applied (both CMake and autotools now pass).

**Comment by mrjimenez** [comment]
> And as you can see there we get a log (96) warnings. Most of them are because of unknown pragma GCC, maybe we should drop these and return to classical #if or #ifdef.

Can you point me to these warnings?

**Comment by Vollstrecker** [comment]
https://github.com/pupnp/pupnp/actions/runs/26589302111/job/78343566202

Seems like I can link to lines in actions output, first one is on line 421 of build-stage. You maybe want to also check lines 37-39 of configure-stage for correctness, and in the install-stage you see why ${SOURCE_DIR} in the genex is required.

**Comment by mrjimenez** [comment]
@Vollstrecker The `#pragma GCC diagnostic` warnings on MSVC (C4068) have been fixed in #590, now merged.

The two affected files were:
- `upnp/src/inc/upnpapi.h` — the `Wunused-function` guard around `GetHandleInfo`
- `upnp/src/urlconfig/urlconfig.c` — the `Wformat-truncation` guard around `snprintf` in `calc_descURL`

Each bare `#pragma GCC diagnostic` block is now wrapped with `#if defined(__GNUC__)` / `#endif`, so MSVC skips them entirely. The Windows CI log for that PR shows zero C4068 hits. Could you confirm the warnings are gone on your end as well?

**Comment by Vollstrecker** [comment]
Perfect. Regarding the change to UPNP_HAS_DEBUG. atm. cmake just does DEBUG/NDEBUG what autotools do, I haven't checked. Should we move to just use the UPNP define or are both for different purposes?

**Comment by mrjimenez** [comment]
@Vollstrecker They serve different purposes and should both exist:

- **`NDEBUG`** is the standard C mechanism. `upnpdebug.h` uses it to replace `UpnpInitLog`, `UpnpSetLogLevel`, `UpnpCloseLog`, and `UpnpGetDebugFile` with inline no-ops — this controls the actual runtime behaviour.
- **`UPNP_HAVE_DEBUG`** is a capability query in `upnpconfig.h` — it lets downstream code and tests ask whether the installed library was built with debug logging support.

The gap you identified is real: CMake never set `UPNP_HAVE_DEBUG`, so `#ifdef UPNP_HAVE_DEBUG` guards in tests always saw the *disabled* path even in Debug builds. PR #592 fixes this by adding an `UPNP_ENABLE_DEBUG` option (default `OFF`, matching the autotools default) and wiring it into `upnpconfig.h`.
